### ISSUE TITLE
Break Application Gateway blocks out into their own resources

### ISSUE THREAD

**Issue by mangoleaf** [hashicorp/terraform-provider-azurerm#23291]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

There are so many blocks that are expected to be defined within this resource and none of them can be composited on to the resource after it has been created without modifying the resource itself.

I am suggesting that we should break these blocks out into their own resources that link to the application gateway id. This would allow the application resource to be created with a lot less dependencies. Furthermore, as resources that need to register with or configure the application gateway are created, they can composite the configuration on at that time.

### New or Affected Resource(s)/Data Source(s)

azurerm_3.73.0

### Potential Terraform Configuration

```hcl
resource "azurerm_application_gateway" "gateway" {
    # base app gateway settings
    
    # all blocks stated as 'one or more' in the documentaiton have been split out into their own resources
}

resource "azurerm_application_gateway_backend_address_pool" {
    application_gateway_id = azurerm_application_gateway.gateway.id
    
    # same backend address pool settings go here
}
```

**Comment by abydal** [comment]
Which issue is this issue duplicate of and what is the upstream MS blocking issue?

**Comment by LairdWilliams** [comment]
I think part of the challenge for this is that AGW deployments in Azure currently (and unfortunately) require at least one backend pool, one listener, one rule, maybe other related bits as well. You can't just set up an "empty" AGW, even through the portal or the APIs and then independently add your first "application" (for lack of a better name for the abstraction represented by an integrated combination of ports, listeners, rules, pools, etc.).

**Comment by mikew3432** [comment]
100% this needs thorough treatment. If truly [Terraform is a first-class citizen in the world of Azure](https://brendanthompson.com/microsoft-and-their-terraform-providers-whats-next/) then Microsoft needs to address this in the back end.

The Terraform code for an Application Gateway of any complexity ends up being a huge monolithic beast.

Terraform plans for even relatively trivial changes to a decent sized configuration - are incomprehensible and dangerous - with huge swathes of "remove" plan followed by huge swathes of "add" of the same things back. At least you HOPE it's the same thing getting added back. There is no good way to be sure you are getting it right, you can only hope your code is good. It's terrifying (no joke).

This is a duplicate of **[#6703 Support for azurerm_application_gateway configuration blocks as separate resources](https://github.com/hashicorp/terraform-provider-azurerm/issues/6703)** from 2020 which admitted defeat fairly quickly... I hope this gets traction.

**Comment by ronmanthe** [comment]
I am with mike on this one. Some sort of effort should be put in here on such an important resource.

**Comment by code-on-git** [comment]
I totally agree with the author. This should definately be fixed in some way.
But when you look att the Terraform documentation about azurerm_application_gateway you can see this:

"The backend_address_pool, backend_http_settings, http_listener, private_link_configuration, request_routing_rule, redirect_configuration, probe, ssl_certificate, and frontend_port properties are Sets as the service API returns these lists of objects in a different order from how the provider sends them. As Sets are stored using a hash, if one value is added or removed from the Set, Terraform considers the entire list of objects changed and the plan shows that it is removing every value in the list and re-adding it with the new information. **Though Terraform is showing all the values being removed and re-added, we are not actually removing anything unless the user specifies a removal in the configfile**."

I did some tests by downloading a large file through the App Gw at the same time I was deploying a change, I noticed that everything works fine during the deploy. But just when the deploy was done, the download session was interrupted.
But the same thing happens when you do updates in the Azure portal.

Is that really by design? Should all sessions be interrupted every time you change something in the App Gw?

**Comment by sluscombepure** [comment]
This would definitely be a welcome addition to the provider. I'm just starting work on a new project and want to add another backend to my existing application gateway, but doing this means I need to edit my TF code in a different repository rather than just import it as a data resource and add onto it.
